tests/functional/test_linear_tracer.py

```diff
@@ -0,0 +1,262 @@
+    return setup.results
+
+
+# First parametrization is over number of cells in pipe and setup for fixture.
+# Second to test all scheduled indices.
+@pytest.mark.parametrize(
+    "results",
+    [
+        (SimplePipe2D.pipe_length / 40, TracerFlowSetup_1p),
+        (SimplePipe2D.pipe_length / 40, TracerFlowSetup_1p_ff),
+    ],
+    indirect=["results"],
+)
+@pytest.mark.parametrize("time_index", [0, 1, 2, 3])
+def test_linear_tracer_1p(time_index: int, results: list[LinearTracerSaveData]) -> None:
+    """Testing the simulation results for the linear tracer with 1-phase.
+
+    Checking that the L2-errors for pressure are small, and that the maximum
+    number of iterations does not exceed 2.
+
+    Checks compares the tracer fraction with an analytical solution accounting
+    for the numerical diffusion of the schemes used (Upwinding & backward Euler).
+
+    """
+
+    sol_data = results[time_index]
+
+    # After the first time step, more iterations are possible because the pressure
+    # must converge to its stationary profile.
+    if time_index == 0:
+        assert sol_data.num_iter <= 5
+    # After pressure converged, linear transport should converge within 1 iteration.
+    # But due to Upwinding, it is sometimes 2.
+    else:
+        assert sol_data.num_iter <= 2
+
+    # Testing errors in pressure (exact).
+    np.testing.assert_allclose(sol_data.error_p, 0.0, atol=1e-7, rtol=0.0)
+
+    # NOTE: Due to the hyperbolic nature, the error in the tracer fraction should
+    # converge to zero linearly, which is checked in a separate test. Here we check only
+    # that the error is not getting worse with ongoing code development.
+    expected_error_z = [
+        0.1128851432293962,
+        0.1344878688028854,
+        0.1489878242519587,
+        0.15299862798949995,
+    ]
+    expected_error_diffused_z = [
+        0.020278908152713,
+        0.018535428299520068,
+        0.01703231859562048,
+        0.016602357208773923,
+    ]
+    # Allow 10% margin for machine reasons.
+    assert (
+        sol_data.error_diffused_z_tracer <= 1.1 * expected_error_diffused_z[time_index]
+    )
+    assert sol_data.error_z_tracer <= 1.1 * expected_error_z[time_index]
+
+
+# Expected to fail because Upwinding and backward euler are diffusive
+@pytest.mark.xfail
+@pytest.mark.parametrize(
+    "results",
+    [
+        (SimplePipe2D.pipe_length / 40, TracerFlowSetup_1p),
+        (SimplePipe2D.pipe_length / 40, TracerFlowSetup_1p_ff),
+    ],
+    indirect=["results"],
+)
+@pytest.mark.parametrize("time_index", [0, 1, 2, 3])
+def test_linear_tracer_1p_is_diffusive(
+    time_index: int, results: list[LinearTracerSaveData]
+) -> None:
+    """Compares the tracer fraction with the exact solution with a steep front.
+
+    This test is expected to fail since the numerical algorithm used is diffusive.
+
+    """
+    sol_data = results[time_index]
+    np.testing.assert_allclose(sol_data.error_z_tracer, 0.0, atol=1e-7, rtol=0.0)
```

Comment:
I suggest we delete this one. There will not be a perfect fit with the analytical solution, unless we implement a front tracking scheme, and computing something and then ignoring  the results is not really meaningful.

Reply:
Deleted.